### PR TITLE
Fix version manager race

### PR DESCRIPTION
**Pull request by marun**
Previously more than one thread could potentially modify VersionManager.versions.  This PR ensures that version writes are
synchronous to avoid that possibility.

### CODE REVIEW

**Comment by marun** [comment]
The hassle of having to use the meta accessor suggests a switch to unstructured.

**Comment by marun** [comment]
Related: #658

**Comment by marun** [comment]
After a conversation with @irfanurrehman, I've come to the conclusion that the current scheme to write versions to the api asynchronously is too complicated.  A switch to synchronous writes to the API is suggested.

**Comment by marun** [comment]
@xunpan @irfanurrehman Updated to write synchronously.  This means that there should never be conflicting access to objects in the map, since only a single thread at a time will be interacting with the version manage for a given key.

**Comment by marun** [comment]
Need to address the namespaced version manager before review.

File: pkg/controller/sync/version/manager.go [hunk 1]
```diff
@@ -276,146 +263,137 @@ func (m *VersionManager) versionQualifiedName(qualifiedName util.QualifiedName)
...
+			// Clear the resource version and retry.
+			resourceVersion = ""
+			return false, nil
+		}
 
-	// Version has been updated or deleted since the last version
-	// manager write.  Attempt to refresh and retry.
+		// Update was successful. All returns should be true even in
+		// the event of an error since the next reconcile can refresh
+		// the resource version on the next reconcile.
```
[the remark this answers is not in the document]
Comment (marun) [reply pkg/controller/sync/version/manager.go]: Done

**Comment by yamt** [comment]
lgtm

**Comment by marun** [comment]
Latest push added a check for errors other than `NotFound` and `Conflict` on update.

File: pkg/controller/sync/version/manager.go [hunk 2]
```diff
@@ -276,146 +263,140 @@ func (m *VersionManager) versionQualifiedName(qualifiedName util.QualifiedName)
...
+
+			glog.V(4).Infof("Creating %s %q", adapterType, qualifiedName)
+			err = m.client.Create(context.TODO(), createdObj)
+			if apierrors.IsAlreadyExists(err) {
+				glog.V(4).Infof("%s %q was created by another process.  Will refresh the resource version and attempt to update.", adapterType, qualifiedName)
+				refreshVersion = true
+				return false, nil
+			}
+			if err != nil {
+				runtime.HandleError(errors.Wrapf(err, "Failed to create version %s %q", adapterType, key))
```
Comment (xunpan) [inline pkg/controller/sync/version/manager.go]: s/version/resource version/

File: pkg/controller/sync/version/manager.go [hunk 3]
```diff
@@ -276,146 +263,140 @@ func (m *VersionManager) versionQualifiedName(qualifiedName util.QualifiedName)
...
-func (m *VersionManager) refreshVersion(obj pkgruntime.Object) error {
-	// A read lock is not suggested due to the name and namespace of a
-	// resource being immutable.
-	qualifiedName := util.NewQualifiedName(obj)
-
-	glog.V(4).Infof("Refreshing %s version %q from the API", m.federatedKind, qualifiedName)
-	refreshedObj := m.adapter.NewObject()
-	err := m.client.Get(context.TODO(), refreshedObj, qualifiedName.Namespace, qualifiedName.Name)
+func (m *VersionManager) getResourceVersionFromAPI(qualifiedName util.QualifiedName) (string, error) {
+	glog.V(4).Infof("Retrieving resourceVersion for %s %q from the API", m.federatedKind, qualifiedName)
```
Comment (xunpan) [inline pkg/controller/sync/version/manager.go]: `s/resourceVersion/resource verion/`

File: pkg/controller/sync/version/manager.go [hunk 4]
```diff
@@ -276,146 +263,140 @@ func (m *VersionManager) versionQualifiedName(qualifiedName util.QualifiedName)
...
+				refreshVersion = true
+				return false, nil
+			}
+			if err != nil {
+				runtime.HandleError(errors.Wrapf(err, "Failed to create version %s %q", adapterType, key))
+				return false, nil
+			}
+
+			// Update the resource version that will be used for update.
+			resourceVersion, err = getResourceVersion(createdObj)
```
Comment (xunpan) [inline pkg/controller/sync/version/manager.go]: For the created resource, it seems no need to update again.
Do you trying to reuse the logic at the end of block to update local cached obj?

Comment (marun) [reply pkg/controller/sync/version/manager.go]: The version detail is stored in the status of the resource.  Status is not set on creation - it needs to be set explicitly through the status subresource after creation.

File: pkg/controller/sync/version/manager.go (hunk 3 above)
Comment (marun) [reply pkg/controller/sync/version/manager.go]: It's a field called resourceVersion.

File: pkg/controller/sync/version/manager.go (hunk 2 above)
Comment (marun) [reply pkg/controller/sync/version/manager.go]: 'version' in this case refers to the version resource - the formatted message would be `Failed to create version PropagatedVersion <myns>/<type>-<myname>`.  Reading that suggests that 'version' should just be removed.

File: pkg/controller/sync/version/manager.go [hunk 5]
```diff
@@ -276,146 +263,140 @@ func (m *VersionManager) versionQualifiedName(qualifiedName util.QualifiedName)
...
+			err := setResourceVersion(createdObj, "")
+			if err != nil {
+				runtime.HandleError(errors.Wrapf(err, "Failed to clear the resource version for %s %q", adapterType, key))
+				return false, nil
+			}
+
+			glog.V(4).Infof("Creating %s %q", adapterType, qualifiedName)
+			err = m.client.Create(context.TODO(), createdObj)
+			if apierrors.IsAlreadyExists(err) {
+				glog.V(4).Infof("%s %q was created by another process.  Will refresh the resource version and attempt to update.", adapterType, qualifiedName)
```
[the remark this answers is not in the document]
Comment (marun) [reply pkg/controller/sync/version/manager.go]: Done

**Comment by marun** [comment]
@xunpan I've updated the log messages for consistency in a new commit - PTAL.

**Comment by xunpan** [comment]
LGTM

**Pull request merged**